### PR TITLE
[CoSim] FEM - PFEM coupling

### PR DESCRIPTION
**Pull request by ceulogiof**
**📝 Description**
This PR introduces the coupling bewteen the `StructuralMechanicsApplication` and the `PfemFluidDynamicsApplication` within the `CoSimulationApplication`. It features several new files in the `PfemFluidDynamicsApplication` for the coupling to work, but with no modifications to the solver files in themselves.

Please feel free to add appropriate labels, as I am not sure which ones to include.

I would also like to have comments on the tests. I provided one example case `pfem_fem_waterslide2d`, but I am not sure if I also need to add individual tests for each coupling operation.

**🆕 Changelog**
- Added the _CouplingOperation_: _ComputeBoundaryForce_. This computes the boundary forces from the PFEM to the FEM using the unit normal of the interface.
- Added _CouplingOperation_: _ResetPfemKinematics_. This resets the PFEM solver kinematic values (node coordinates, displacement, velocity and acceleration) back to its last converged state. It is adapted from the `FemToDemApplication`. The new C++ files are included as a PFEM custom utility.
- Added _CouplingOperation_: _PrintIterationNumberOperation_. This is mentioned in #5140. 
- Added the Fix/Free velocity process (adapted from the `FemToDemApplication`) in the PFEM Solver Wrapper. The new C++ files are included as a PFEM custom process.

### CODE REVIEW

**Comment by AlejandroCornejo** [comment]
Overall looks fine to me! I also know the work behind this and I know everything is correct! Even though you do not modify anything inside PFEM I tag @AFranci to be aware of the PFEM additions. Good work!

**Comment by philbucher** [comment]
Cool!
I will review in the next days

File: applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                if(self.write_output_file):
+                    self.output_file.write(' '.join(output_values) + '\n')
+    
+    def _EvaluateGlobalForces(self):
+        # vel_x, vel_y, vel_z
+        velocity = [0.0, 0.0, 0.0]
+        sum_forces = [0.0, 0.0, 0.0]
+        pressure_list = [0.0]
+
+        # aktuellen Zeitstep zurück setzen - warum ist das nötig ?
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py]: nice :)

File: applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        # TODO !!
+        for node in self.model_part.Nodes:
+            node.SetSolutionStepValue(KM.REACTION_X, 0, 0)
+            node.SetSolutionStepValue(KM.REACTION_Y, 0, 0)
+            node.SetSolutionStepValue(KM.REACTION_Z, 0, 0)
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py]: you should use the `VariableUtils` for this. I think the fct is `SetVariable`

File: applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    def _PrintUnitNormalInfo(self, nodes, element, unit_normal):
+        # print normal debug information
+        import pdb
+        pdb.set_trace()
+        cs_tools.cs_print_info('Element info', element)
+        cs_tools.cs_print_info('Unit normal', unit_normal)
+        node_counter = 0
+        for node in nodes:
+            cs_tools.cs_print_info(str(node_counter) + ' Node Id', str(node.Id))
+            node_counter += 1
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py]: this should not go to master as it is only for debuggign

File: applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            self.width = 1
+        
+        # Detect 'End' as a tag and replace it by a large number
+        if(self.settings.Has('interval')):
+            if(self.settings['interval'][1].IsString()):
+                if(self.settings['interval'][1].GetString() == 'End' or self.settings['interval'][1].GetString() == 'end'):
+                    self.settings['interval'][1].SetDouble(1e30)
+                else:
+                    raise Exception('The second value of interval can be \'End\' or a number, interval currently:' + self.settings['interval'].PrettyPrintJsonString())
+        self.interval = self.settings["interval"].GetVector()
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py]: please use the `IntervalUtility` for this

File: applications/CoSimulationApplication/python_scripts/coupling_operations/reset_pfem_kinematics.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        self.model_part = self.model[self.model_part_name]
+
+        # Detect 'End' as a tag and replace it by a large number
+        if(self.settings.Has('interval')):
+            if(self.settings['interval'][1].IsString()):
+                if(self.settings['interval'][1].GetString() == 'End' or self.settings['interval'][1].GetString() == 'end'):
+                    self.settings['interval'][1].SetDouble(1e30)
+                else:
+                    raise Exception('The second value of interval can be \'End\' or a number, interval currently:' + self.settings['interval'].PrettyPrintJsonString())
+        self.interval = self.settings["interval"].GetVector()
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/reset_pfem_kinematics.py]: please use `IntervalUtility`

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp [hunk 6]
```diff
@@ -442,6 +443,41 @@ void TrussElement3D2N::CalculateOnIntegrationPoints(
     if (rOutput.size() != integration_points.size()) {
         rOutput.resize(integration_points.size());
     }
+
+    if (rVariable == PRESTRESS_VECTOR){
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: can you please make a separate PR to the StructuralMechanicsApp and add a test?

File: applications/PfemFluidDynamicsApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    if (mFreeOrFix == 0) 
+    {
+        KRATOS_INFO("FixFreeVelocityOnNodesProcess") << " FIXED VELOCITY VALUES FOR THE NODES " << std::endl;
+    }
+    else
+    {
+        KRATOS_INFO("FixFreeVelocityOnNodesProcess") << " FREED VELOCITY VALUES FOR THE NODES " << std::endl;
+    }
+}
```
Comment (philbucher) [inline applications/PfemFluidDynamicsApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]: I recommend using https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/variable_utils.h#L992-L1025

It does loop the nodes 3 times, but I think it is better to use the default implementation

Alternatively if you want to keep the current implementation then please use the `ParallelUtilities` => search for `block_for_each` in the code, you will get lots of examples

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        {
+            auto it_node = rNodes.begin() + i;
+
+            noalias(it_node->Coordinates()) = it_node->GetInitialPosition().Coordinates();
+            noalias(it_node->Coordinates()) += it_node->FastGetSolutionStepValue(DISPLACEMENT);
+        }
+
+        KRATOS_INFO("MoveMeshUtility") << " PFEM MESH MOVED " << std::endl;
+        KRATOS_CATCH("")
+    }
```
Comment (philbucher) [inline applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp]: same as `VariableUtils::UpdateCurrentPosition`

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    void MoveMeshUtility::ResetPfemKinematicValues(ModelPart& rFluidModelPart)
+    {
+        const auto it_node_begin = rFluidModelPart.NodesBegin();
+
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(rFluidModelPart.Nodes().size()); ++i) {
```
Comment (philbucher) [inline applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp]: `block_for_each`

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    class  KRATOS_API(PFEM_FLUID_DYNAMICS_APPLICATION) MoveMeshUtility
+    {
+        public:
+            typedef ModelPart::NodesContainerType NodesContainerType;
+
+            KRATOS_CLASS_POINTER_DEFINITION(MoveMeshUtility);
+
+            MoveMeshUtility() {};
+            ~MoveMeshUtility() {};
```
Comment (philbucher) [inline applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h]:
```suggestion
            MoveMeshUtility() = delete;
            ~MoveMeshUtility() = default;
```

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+             * @param rNodes The nodes array of a model part
+             */
+            void MovePfemMesh(NodesContainerType& rNodes) const;
+
+            /**
+             * @brief This function resets the previous kinematic
+             *        information of the PFEM nodes.
+             * @param rFluidModelPart The fluid model part
+             */
+            void ResetPfemKinematicValues(ModelPart& rFluidModelPart);
```
Comment (philbucher) [inline applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h]:
```suggestion
            static void ResetPfemKinematicValues(ModelPart& rFluidModelPart);
```

this class should consist of `static` methods only and not have a Constructor defined

File: applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 12]
```diff
@@ -42,6 +43,11 @@ void AddCustomUtilitiesToPython(pybind11::module &m)
                       MultiFileFlag,
                       WriteDeformedMeshFlag,
                       WriteConditionsFlag>());
+    
+    py::class_<MoveMeshUtility, MoveMeshUtility::Pointer>(m, "MoveMeshUtility")
+        .def(py::init<>())
+        .def("MovePfemMesh", &MoveMeshUtility::MovePfemMesh)
+        .def("ResetPfemKinematicValues", &MoveMeshUtility::ResetPfemKinematicValues);
```
Comment (philbucher) [inline applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp]:
```suggestion
        .def_static("ResetPfemKinematicValues", &MoveMeshUtility::ResetPfemKinematicValues);
```

**Review by philbucher — CHANGES_REQUESTED** [review]
Overall nice work :+1: 
I have some comments / suggestions regarding the implementation

Also the test you added in the CoSimApp is not registered so it will not get executed when running all tests. How long does it take? Depending on this I will tell you where to add it

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h (hunk 10 above)
Comment (ceulogiof) [reply applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h]: I tried this (with the `static` modification you suggest below in both the header and python binding), but I get this compilation error, which makes me think that pybind11 tries to use the deleted constructor. I assume you suggested this to prevent the usage of this class without passing any arguments to it. To circumvent this, I did `MoveMeshUtility() = default`, which compiles but this replicates the older behavior. Do you have any other hint about this error?


```
In file included from /home/sose2021/Software/Kratos/external_libraries/pybind11/pybind11.h:48,
                 from /home/sose2021/Software/Kratos/applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.h:14,
                 from /home/sose2021/Software/Kratos/applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp:15:
/home/sose2021/Software/Kratos/external_libraries/pybind11/detail/init.h: In instantiation of ‘Class* pybind11::detail::initimpl::construct_or_initialize(Args&& ...) [with Class = Kratos::MoveMeshUtility; Args = {}; typename std::enable_if<(! std::is_constructible<_Tp, _Args>::value), int>::type <anonymous> = 0]’:
/home/sose2021/Software/Kratos/external_libraries/pybind11/detail/init.h:175:66:   required from ‘static void pybind11::detail::initimpl::constructor<Args>::execute(Class&, const Extra& ...) [with Class = pybind11::class_<Kratos::MoveMeshUtility, std::shared_ptr<Kratos::MoveMeshUtility> >; Extra = {}; typename std::enable_if<(! Class::has_alias), int>::type <anonymous> = 0; Args = {}]’
/home/sose2021/Software/Kratos/external_libraries/pybind11/pybind11.h:1370:9:   required from ‘pybind11::class_<type_, options>& pybind11::class_<type_, options>::def(const pybind11::detail::initimpl::constructor<Args ...>&, const Extra& ...) [with Args = {}; Extra = {}; type_ = Kratos::MoveMeshUtility; options = {std::shared_ptr<Kratos::MoveMeshUtility>}]’
/home/sose2021/Software/Kratos/applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp:48:26:   required from here
/home/sose2021/Software/Kratos/external_libraries/pybind11/detail/init.h:63:64: error: use of deleted function ‘Kratos::MoveMeshUtility::MoveMeshUtility()’
   63 | inline Class *construct_or_initialize(Args &&...args) { return new Class{std::forward<Args>(args)...}; }
      |                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/sose2021/Software/Kratos/applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp:25:
/home/sose2021/Software/Kratos/applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h:49:13: note: declared here
   49 |             MoveMeshUtility() = delete;
      |             ^~~~~~~~~~~~~~~
make[2]: *** [applications/PfemFluidDynamicsApplication/CMakeFiles/KratosPfemFluidDynamicsApplication.dir/build.make:102: applications/PfemFluidDynamicsApplication/CMakeFiles/KratosPfemFluidDynamicsApplication.dir/custom_python/add_custom_utilities_to_python.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2516: applications/PfemFluidDynamicsApplication/CMakeFiles/KratosPfemFluidDynamicsApplication.dir/all] Error 2
make: *** [Makefile:130: all] Error 2

```

File: applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py (hunk 1 above)
Comment (ceulogiof) [reply applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py]: The test takes 1 minute 55 seconds on the remote PC I have assigned. I guess it should go onto the nightly, right?

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp (hunk 6 above)
Comment (ceulogiof) [reply applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: @KlausBSautter Did you add this to be able to have a pre-stress on the membrane for the water slide case in 3D?

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h (hunk 10 above)
Comment (ceulogiof) [reply applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h]: It is also worth noting that I removed the `MoveMeshUtility::MovePfemMesh` function as it is already implemented in the `VariableUtils::UpdateCurrentPosition`, and I don't use it in the coupling (currently I only use the function `ResetPfemKinematicValues`). After that removal and the suggested edits the above compilation error happened.

Comment (philbucher) [reply applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h]: Can you also my suggestion in the python interface?
Then this error should be gone

File: applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 12 above)
Comment (philbucher) [reply applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp]: This one I mean

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h [hunk 13]
```diff
@@ -34,38 +34,28 @@ namespace  Kratos
...
-            MoveMeshUtility() {};
-            ~MoveMeshUtility() {};
-
-            /**
-             * @brief This function is the main operation of this utility.
-             *        It moves the nodes of a given model part w.r.t. the current DISPLACEMENT
-             * @param rNodes The nodes array of a model part
-             */
-            void MovePfemMesh(NodesContainerType& rNodes) const;
+            MoveMeshUtility() = default;
```
Comment (ceulogiof) [inline applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.h]: @philbucher I still have the same issue as I mentioned [here](https://github.com/KratosMultiphysics/Kratos/pull/9294#discussion_r752482809), the code doesn't compile if I delete the constructor here. I did change the binding to be static as well.

File: applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 14]
```diff
@@ -46,8 +46,7 @@ void AddCustomUtilitiesToPython(pybind11::module &m)
     
     py::class_<MoveMeshUtility, MoveMeshUtility::Pointer>(m, "MoveMeshUtility")
         .def(py::init<>())
-        .def("MovePfemMesh", &MoveMeshUtility::MovePfemMesh)
-        .def("ResetPfemKinematicValues", &MoveMeshUtility::ResetPfemKinematicValues);
+        .def_static("ResetPfemKinematicValues", &MoveMeshUtility::ResetPfemKinematicValues);
```
Comment (ceulogiof) [inline applications/PfemFluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp:49]: @philbucher Here's the static definition in the Python binding.

File: applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py (hunk 2 above)
Comment (ceulogiof) [reply applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py]: I addressed this, but later in:

https://github.com/KratosMultiphysics/Kratos/blob/f6204c015c6a79d05006d3d5a8d814161740d023/applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py#L104-L106

I left them unchanged because I compute some values elementwise, and then apply them to the nodes.

File: applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py (hunk 3 above)
Comment (ceulogiof) [reply applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py]: I eliminated this now.

File: applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py (hunk 4 above)
Comment (ceulogiof) [reply applications/CoSimulationApplication/python_scripts/coupling_operations/compute_boundary_force.py]: Thanks for this hint, it reduced all of this to a single line.

File: applications/CoSimulationApplication/python_scripts/coupling_operations/reset_pfem_kinematics.py (hunk 5 above)
Comment (ceulogiof) [reply applications/CoSimulationApplication/python_scripts/coupling_operations/reset_pfem_kinematics.py]: I now changed it.

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp (hunk 6 above)
Comment (ceulogiof) [reply applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: I eliminated it definitively, because I talked with Klaus and this was no longer required, nor used in the PFEM coupling.

File: applications/PfemFluidDynamicsApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp (hunk 7 above)
Comment (ceulogiof) [reply applications/PfemFluidDynamicsApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]: I saw the examples and now changed it accordingly.

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp (hunk 8 above)
Comment (ceulogiof) [reply applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp]: I now eliminated it entirely.

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp (hunk 9 above)
Comment (ceulogiof) [reply applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp]: I addressed this now.

File: applications/PfemFluidDynamicsApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    VariableUtils().ApplyFixity(VELOCITY_Z, mFreeOrFix, mrModelPart.Nodes());
+
+    if (mFreeOrFix) 
+    {
+        KRATOS_INFO("FixFreeVelocityOnNodesProcess") << " FIXED VELOCITY VALUES FOR THE NODES " << std::endl;
+    }
+    else
+    {
+        KRATOS_INFO("FixFreeVelocityOnNodesProcess") << " FREED VELOCITY VALUES FOR THE NODES " << std::endl;
+    }
```
Comment (philbucher) [inline applications/PfemFluidDynamicsApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]: I suggest to wrap this in an echo level as it can be very verbose

File: applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+            noalias(rNode.Coordinates()) = rNode.GetInitialPosition() + old_disp;
+
+            noalias(r_current_disp) = old_disp;
+            noalias(r_current_vel)  = rNode.FastGetSolutionStepValue(VELOCITY, 1);
+            noalias(r_current_acc)  = rNode.FastGetSolutionStepValue(ACCELERATION, 1);
+        }
+    });
+
+    KRATOS_INFO("MoveMeshUtility::ResetPfemKinematicValues") << " PFEM KINEMATICS RESET " << std::endl;
```
Comment (philbucher) [inline applications/PfemFluidDynamicsApplication/custom_utilities/move_mesh_utility.cpp]: same here it would be best to wrap this in an echo level

**Review by philbucher — COMMENTED** [review]
@ceulogiof I did some fixes, now also the static should work. Please check your tests, but it should be ok. I have two more final comments:
- Please wrap the logger calls in echo levels, otherwise it is very verbose
- Please add the test [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CoSimulationApplication/tests/co_simulation_test_factory.py) so that it gets executed when running the main test script

**Comment by ceulogiof** [comment]
Hey @philbucher I wrapped the messages into the echo level 1, but decided to move this into the python scripts rather than in the cpp code. I hope this is fine, if not let me know.
As per the test, I have a couple of issues that I need some help with. I will leave the specific comments in the files.

File: applications/CoSimulationApplication/tests/co_simulation_test_factory.py [hunk 17]
```diff
@@ -205,6 +206,16 @@ def test_sdof_fsi(self):
...
+    def test_PFEM_FEM_water_slide_2d(self):
+        if not numpy_available:
+            self.skipTest("Numpy not available")
+        if not have_pfem_fem_dependencies:
+            self.skipTest("PFEM FEM dependencies are not available!")
+
+        with KratosUnittest.WorkFolderScope(".", __file__):
+            self._createTest("pfem_fem_waterslide2d","cosim_pfem_fem_waterslide2d")
+            self._runTest()
+
```
Comment (ceulogiof) [inline applications/CoSimulationApplication/tests/co_simulation_test_factory.py:225]: There are two leftover files when I run the test locally in my machine, producing the following output:

```
/home/sose2021/Software/Kratos/applications/CoSimulationApplication/tests/co_simulation_test_case.py:33: ResourceWarning: unclosed file <_io.TextIOWrapper name='totalVolumeBeforeMeshing.txt' mode='w' encoding='UTF-8'>
  CoSimulationAnalysis(self.cosim_parameters).Run()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
/home/sose2021/Software/Kratos/applications/CoSimulationApplication/tests/co_simulation_test_case.py:33: ResourceWarning: unclosed file <_io.TextIOWrapper name='PfemFluidModelPart.Fluid_number_iterations.dat' mode='w' encoding='UTF-8'>
  CoSimulationAnalysis(self.cosim_parameters).Run()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
ok
```
One of these files is the number of iterations from the print_iteration_number coupling operation. I can remove this from the co_sim_parameters... but I thought the test could take care of this.
The other file is always produced when running PFEM simulations, so there I am also not sure how to proceed.
Am I supposed to manually erase these files with the `tearDownClass()` at the bottom of the test factory?
Also, I put my test into the `TestCoSimulationCases`, because if I tried to put it locally into the nightly, I couldn't make it run.

Finally, there is a lot of output produced when running the test (locally), but only due to the PFEM solver. I don't know why, because I quickly checked the message output in the solver and all of it is also wrapped to an echo level.

Comment (philbucher) [reply applications/CoSimulationApplication/tests/co_simulation_test_factory.py:225]: I fixed it, basically I disabled all the output
Unless necessary for the test we do not write any output

One final concern: There are many warnings (I think from the structural solver), can you please check and fix them
~~~
[WARNING] ResidualBasedBlockBuilderAndSolver: ATTENTION! setting the RHS to zero!
[WARNING] DisplacementCriteria: NaN norm is detected. Setting reference to convergence criteria
~~~

Comment (ceulogiof) [reply applications/CoSimulationApplication/tests/co_simulation_test_factory.py:225]: @philbucher I addressed this and now shouldn't have these warning messages. I changed the parameter files, not the code itself.

**Comment by ceulogiof** [comment]
@philbucher Let me know if there are any other issues with the code, in case something slipped until this point.

**Review by philbucher — APPROVED** [review]
awesome, now it is perfect, thx for the patience!

**Comment by philbucher** [comment]
@mpentek @RiedlAndreas maybe this is sth you are also interested... (you use the PfemFluidApp right?)

**Comment by ceulogiof** [comment]
@philbucher thanks you for all your support and patience with my code as well!
